pkg/ddevapp/commands.go

```diff
@@ -1,47 +1,69 @@
+
+	dockerImages "github.com/ddev/ddev/pkg/docker"
+	"github.com/ddev/ddev/pkg/dockerutil"
+	"github.com/ddev/ddev/pkg/globalconfig"
+	"github.com/ddev/ddev/pkg/nodeps"
+	"github.com/ddev/ddev/pkg/util"
 )
 
-// PopulateCustomCommandFiles sets up the custom command files in the project
+// PopulateGlobalCustomCommandFiles sets up the custom command files in the project
 // directories where they need to go.
-func PopulateCustomCommandFiles(app *DdevApp) error {
-
+func PopulateGlobalCustomCommandFiles() error {
 	sourceGlobalCommandPath := filepath.Join(globalconfig.GetGlobalDdevDir(), "commands")
 	err := os.MkdirAll(sourceGlobalCommandPath, 0755)
 	if err != nil {
 		return nil
 	}
 
-	projectCommandPath := app.GetConfigPath("commands")
-	// Make sure our target global command directory is empty
-	copiedGlobalCommandPath := app.GetConfigPath(".global_commands")
-	err = os.MkdirAll(copiedGlobalCommandPath, 0755)
+	// Remove contents of the directory, if the directory exists and has some contents
+	commandDirInVolume := "/mnt/ddev-global-cache/global-commands/"
+	_, stderr, err := performTaskInContainer([]string{"sh", "-c", "rm -r " + path.Join(commandDirInVolume, "*") + " || true"})
 	if err != nil {
-		util.Error("Unable to create directory %s: %v", copiedGlobalCommandPath, err)
-		return nil
+		return fmt.Errorf("unable to rm %s: %v (stderr=%s)", commandDirInVolume, err, stderr)
 	}
 
-	// Make sure it's empty
-	err = fileutil.PurgeDirectory(copiedGlobalCommandPath)
+	// Copy commands into container (this will create the directory if it's not there already)
+	uid, _, _ := util.GetContainerUIDGid()
+	err = dockerutil.CopyIntoVolume(sourceGlobalCommandPath, "ddev-global-cache", "global-commands", uid, "host", false)
 	if err != nil {
-		util.Error("Unable to remove %s: %v", copiedGlobalCommandPath, err)
-		return nil
+		return err
 	}
 
-	err = copy2.Copy(sourceGlobalCommandPath, copiedGlobalCommandPath)
+	// Make sure all commands can be executed
+	_, stderr, err = performTaskInContainer([]string{"sh", "-c", "chmod -R u+rwx " + commandDirInVolume})
 	if err != nil {
-		return err
+		return fmt.Errorf("unable to chmod %s: %v (stderr=%s)", commandDirInVolume, err, stderr)
 	}
 
-	if !fileutil.FileExists(projectCommandPath) || !fileutil.IsDirectory(projectCommandPath) {
-		return nil
-	}
 	return nil
 }
+
+// performTaskInContainer runs a command in the web container if it's available,
+// but uses an anonymous container otherwise.
+func performTaskInContainer(command []string) (string, string, error) {
+	app, err := GetActiveApp("")
+	if err == nil {
+		status, _ := app.SiteStatus()
+		if status == SiteRunning {
+			// Prepare docker exec command
+			opts := &ExecOpts{
+				RawCmd:    command,
+				Tty:       false,
+				NoCapture: false,
+			}
+			return app.Exec(opts)
+		}
+	}
+
+	// If there is no running active site, use an anonymous container instead.
+	containerName := "PopulateGlobalCustomCommandFiles" + nodeps.RandomString(12)
```

Comment:
Refactored function (thanks!) didn't get updated anon name. Doesn't matter but might confuse some poor dev someday.
```suggestion
	containerName := "performTaskInContainer" + nodeps.RandomString(12)
```

Reply:
It's only ever used from `PopulateGlobalCustomCommandFiles` which is why I kept the name how it was. But I'll make this requested change.

Reply:
Done